ODIN_II/SRC/odin_error.cpp

```diff
@@ -107,8 +106,29 @@ void _log_message(odin_error error_type, long column, long line_number, long fil
             fprintf(stderr, "\n");
     }
 
-    print_culprit_line(column, line_number, file);
+    if (file >= 0 && (size_t)file < include_file_names.size()
+        && line_number >= 0) {
+        print_culprit_line(column, line_number, include_file_names[file].first.c_str());
+    }
 
     fflush(stderr);
-    _verbose_assert(!fatal_error, "", function_file_name, function_line, function_name);
+    _verbose_assert(!fatal_error, NULL, function_file_name, function_line, function_name);
+}
+
+void _verbose_assert(bool condition, const char* condition_str, const char* odin_file_name, long odin_line_number, const char* odin_function_name) {
+    fflush(stdout);
+    if (!condition) {
+        fprintf(stderr, "%s:%ld: %s: ", odin_file_name, odin_line_number, odin_function_name);
+        if (condition_str) {
+            // We are an assertion, print the condition that failed and which line it occurred on
+            fprintf(stderr, "Assertion %s failed\n", condition_str);
+            // -1 since __LINE__ starts from 1
+            print_culprit_line(-1, odin_line_number - 1, odin_file_name);
+        } else {
+            // We are a parsing error, dont print the culprit line
+            fprintf(stderr, "Fatal error\n");
```

Comment:
Is there any value to printing the line in odin for this case? Seems like the user shouldn't care and devs should be able to just look at the code.

Reply:
I don't disagree, the biggest advantage here is when a user report an error, now we can actually see what failed until a better issue is reported. More often than not all that is sent is just the error message, no test file etc... so it's hard to reproduce and this gives us some bread crumbs. While I agree in the long term this should go away, right now I don't know.

Reply:
What I meant is should we be printing the culprit line instead of just the line number. I definitely see the value of keeping the ODIN line in for the foreseeable future. But I think printing the actual code in ODIN will be confusing for the user and should not be relevant to them, since (in theory) its a bug in their code not ours.
So for now we will continue to print the ODIN line number and no extra information

Reply:
My bad. I thought you meant line number :face-palm:.
Then no, line number is all the info we want here. I agree that it would just be confusing.